### PR TITLE
Fake checkins - move business logic to submission service

### PR DESCRIPTION
**Pull request by alexmoroianu**
EXPOSUREBACK-1091

- Move fake checkins generation on submission service from persistance modules.
- Create fake checkins facade to hide the complexity of saving fake checkins in the Submission Controller.
- Refactor duplicated code and centralize it.

### CODE REVIEW

**Review by ioangut — APPROVED** [review]

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/service/TraceTimeIntervalWarningService.java [hunk 1]
```diff
@@ -33,22 +31,17 @@
...
-   *                                     data.
    * @throws NoSuchAlgorithmException In case the MessageDigest used in hashing can not be instantiated.
    */
   public TraceTimeIntervalWarningService(
-      TraceTimeIntervalWarningRepository traceTimeIntervalWarningRepo,
-      FakeCheckinsGenerator fakeCheckinsGenerator) throws NoSuchAlgorithmException {
+      TraceTimeIntervalWarningRepository traceTimeIntervalWarningRepo) throws NoSuchAlgorithmException {
     this.traceTimeIntervalWarningRepo = traceTimeIntervalWarningRepo;
-    this.fakeCheckinsGenerator = fakeCheckinsGenerator;
     this.hashAlgorithm = MessageDigest.getInstance("SHA-256");
```
Comment (FelixRottler) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/service/TraceTimeIntervalWarningService.java:45]: Maybe replace this hardcoded SHA-256 with the new ENUM you created?

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  /**
+   * Generates a random byte array data for a specific size.
+   *
+   * @param size - byte array size
+   * @return - random key
+   */
+  public static byte[] generateRandomByteArrayData(int size) {
+    byte[] randomKeyData = new byte[size];
+    new SecureRandom().nextBytes(randomKeyData);
```
Comment (FelixRottler) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java:37]: Maybe we need to think about security here. A SecureRandom does not produce the same random number when using the same seed. Does this provide an advantage?

File: common/persistence/src/test/java/app/coronawarn/server/common/persistence/utils/hash/HashUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package app.coronawarn.server.common.persistence.utils.hash;
+
+import app.coronawarn.server.common.persistence.utils.hash.HashUtils.MessageDigestAlgorithms;
+import com.google.protobuf.ByteString;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class HashUtilsTest {
```
Comment (FelixRottler) [inline common/persistence/src/test/java/app/coronawarn/server/common/persistence/utils/hash/HashUtilsTest.java:9]: The test is missing some scenarios for "byteStringDigest". Maybe we can add an additional test to cover that?

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    return bytesToHex(md.digest());
+  }
+
+  /**
+   * Transform bytes array into HEX.
+   * @param bytes - array to be transformed
+   * @return - Hex value
+   */
+  public static String bytesToHex(byte[] bytes) {
+    StringBuffer result = new StringBuffer();
```
Comment (FelixRottler) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java:86]: Missing test for this method

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  /**
+   * Hash of the provided string with wanted algorithm and return it as HEX.
+   * @param data - string to be hash
+   * @param algorithm - algorithm to be used
+   * @return - HEX string
+   * @throws NoSuchAlgorithmException - thrown when the algorithm does not exist.
+   */
+  public static String digestAsHex(String data, MessageDigestAlgorithms algorithm) throws NoSuchAlgorithmException {
+    MessageDigest md = MessageDigest.getInstance(algorithm.getName());
```
Comment (FelixRottler) [inline common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java:75]: Missing test for this method

File: common/persistence/src/test/java/app/coronawarn/server/common/persistence/TestApplication.java [hunk 6]
```diff
@@ -15,14 +15,14 @@
...
 import app.coronawarn.server.common.persistence.service.common.ValidDiagnosisKeyFilter;
-import app.coronawarn.server.common.persistence.service.utils.checkins.FakeCheckinsGenerator;
 import app.coronawarn.server.common.persistence.utils.YamlPropertySourceFactory;
 import java.security.NoSuchAlgorithmException;
 import java.util.Map;
 import org.mockito.Mockito;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
```
Comment (FelixRottler) [inline common/persistence/src/test/java/app/coronawarn/server/common/persistence/TestApplication.java:25]: Primary is unused import here

File: services/submission/src/main/java/app/coronawarn/server/services/submission/checkins/EventCheckinFacade.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        numberOfFakesToCreate, pepper));
+    return traceTimeIntervalWarningService.saveCheckins(allCheckins, submissionTimestamp, submissionType);
+  }
+
+  /**
+   * Extract check-ins from submission payload and store them.
+   * @param submissionPayload - submission payload
+   * @return - storage result containing number of filtered and saved check-ins.
+   */
+  public CheckinsStorageResult extractAndStoreEventCheckins(SubmissionPayload submissionPayload) {
```
Comment (FelixRottler) [inline services/submission/src/main/java/app/coronawarn/server/services/submission/checkins/EventCheckinFacade.java:65]: Please make sure to test this method too :)

File: services/submission/src/test/java/app/coronawarn/server/services/submission/controller/FakeRequestControllerTest.java [hunk 8]
```diff
@@ -17,6 +17,7 @@
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Import;
```
Comment (FelixRottler) [inline services/submission/src/test/java/app/coronawarn/server/services/submission/controller/FakeRequestControllerTest.java:20]: This is an unused Import when looking at the file

**Review by FelixRottler — CHANGES_REQUESTED** [review]
See comments

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java (hunk 2 above)
Comment (alexmoroianu) [reply common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java:37]: The previous calls to this generation were used exactly like this. I just moved the code in this method to be re-usable. I changed the name of it though to reflect the secure random.

File: common/persistence/src/test/java/app/coronawarn/server/common/persistence/utils/hash/HashUtilsTest.java (hunk 3 above)
Comment (alexmoroianu) [reply common/persistence/src/test/java/app/coronawarn/server/common/persistence/utils/hash/HashUtilsTest.java:9]: I will add these tests in the following PR regarding shared common module as both Utils and Test class will be re-located.

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java (hunk 4 above)
Comment (alexmoroianu) [reply common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java]: I will add these tests in the following PR regarding shared common module as both Utils and Test class will be re-located.

File: common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java (hunk 5 above)
Comment (alexmoroianu) [reply common/persistence/src/main/java/app/coronawarn/server/common/persistence/utils/hash/HashUtils.java]: Method removed as it was not used.

File: services/submission/src/main/java/app/coronawarn/server/services/submission/checkins/EventCheckinFacade.java (hunk 7 above)
Comment (alexmoroianu) [reply services/submission/src/main/java/app/coronawarn/server/services/submission/checkins/EventCheckinFacade.java:65]: Added.

**Review by FelixRottler — APPROVED** [review]

**Pull request merged**